### PR TITLE
Add neural score estimation (NSE)

### PR DESCRIPTION
**Pull request by francois-rozet**
Targets #6

### CODE REVIEW

File: lampe/inference.py [hunk 1]
```diff
@@ -524,14 +540,183 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
...
+
+    def alpha(self, t: Tensor) -> Tensor:
+        return torch.exp(-8.0 * t**2)
+
+    def beta(self, t: Tensor) -> Tensor:
+        return 16.0 * t
+
+    def ode(self, theta: Tensor, x: Tensor, t: Tensor) -> Tensor:
+        drift = -self.beta(t) / 2 * theta
+        diffusion = -self.beta(t) / 2 * (1 + self.alpha(t)) * self.forward(theta, x, t)
```
Comment (glouppe) [inline lampe/inference.py:648]: Shouldn't we have an additional `(1 - self.alpha(t))` factor here?

File: lampe/inference.py [hunk 2]
```diff
@@ -524,14 +540,183 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
...
+        batch_shape = x.shape[:-1]
+
+        return NormalizingFlow(
+            transform=FFJTransform(
+                f=lambda theta, t: self.ode(theta, x, t),
+                time=x.new_tensor(1.0),
+                phi=(x, *self.parameters()),
+            ),
+            base=DiagNormal(self.zeros, self.ones).expand(batch_shape),
+        )
```
Comment (glouppe) [inline lampe/inference.py:676]: I am not sure if there is any definite API yet but I find a bit inconsistent that `NPE` defines a `sample` method based on its own `self.flow` and here there is no `sample` method implemented.

**Review by glouppe — COMMENTED** [review]
Hi! I have yet to run and play with the code, but here are some comments from just reading the code.

File: lampe/inference.py [hunk 3]
```diff
@@ -524,14 +540,183 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
...
+
+    .. math:: \mathrm{d} \theta = \left[ -\frac{1}{2} \beta(t) \theta -
+        \frac{1}{2} \beta(t) (1 + \alpha(t)) s_\phi(\theta, x, t) \right] \, \mathrm{d} t
+
+    whose trajectories share the same marginal probability densities :math:`p(\theta_t | x)`
+    as the stochastic diffusion process.
+
+    References:
+        | Score-Based Generative Modeling through Stochastic Differential Equations (Song et al., 2021)
+        | https://arxiv.org/abs/2011.13456
```
[the remark this answers is not in the document]
Comment (francois-rozet) [reply lampe/inference.py:588]: These papers are related, but are too different to be references for the implementation, which is solely inspired by Song et al. (2021) who formulate the forward diffusion process as an SDE.

The main contribution of Geffner et al. (2022) is to show that a one-observation score-based posterior estimator can be used to sample from the many-observations posterior. They use an older training approach which discretizes the diffusion process and doesn't reweight the terms at different timesteps `t`, which is now standard. The sampling method is also custom.

Sharrock et al. (which isn't published yet) use the same SDE formulation as Song et al. (2021), but still discretize the diffusion process for training (and probably sampling, although it is never mentioned how sampling is performed). Additionally, they use a variance exploding SDE as diffusion process, which is known to produce much worse results than variance preserving ones. Finally, the main contribution of the paper is to extend score-matching to the sequential setting.

Most importantly, log-density computations are not tackled by either of the papers.

File: lampe/inference.py (hunk 1 above)
Comment (francois-rozet) [reply lampe/inference.py:648]: I don't think so. I used a rescaled version (of my own) of the score-matching objective such that

$$s_\phi(\theta, x, t) \approx (1 - \alpha(t)) \nabla_\theta ~ p_t(\theta | x)$$

Thanks to that the variance of the estimator is uniform with respect to the time, which makes the training and ODE more stable (at least in my experiments).

File: lampe/inference.py (hunk 2 above)
Comment (francois-rozet) [reply lampe/inference.py:676]: Yes, I dislike it as well, but I am not sure of the best way to sample from the trained estimator yet. From my toy experiments sampling from the probability flow ODE gives good samples and is pretty fast, but Song et al. (2021) put forward some evidence that ancestral and reverse SDE sampling yields more qualitative samples.

File: lampe/inference.py [hunk 4]
```diff
@@ -524,14 +540,183 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
...
 
         return -log_prob.mean()
 
 
+class NSE(nn.Module):
+    r"""Creates a neural score estimation (NSE) regression network.
+
+    The principle of neural score estimation is to train a regression network
+    :math:`s_\phi(\theta, x, t)` to approximate the score of a variance preserving
+    stochastic diffusion process
```
[the remark this answers is not in the document]
Comment (francois-rozet) [reply lampe/inference.py:556]: Using the "sub-VP SDE" name requires "VP SDE" to be introduced first and maybe a comparison, which I think would be confusing.

Besides, I don't like the name "VP SDE" for the DDPM-style diffusion process as it is not the only variance preserving one. In fact the "sub-VP SDE" is also variance preserving, it just has a lower variance than the "VP SDE".

File: lampe/inference.py [hunk 5]
```diff
@@ -524,14 +540,183 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
...
+    ):
+        super().__init__()
+
+        if moments is None:
+            self.standardize = nn.Identity()
+        else:
+            mu, sigma = moments
+            self.standardize = Affine(-mu / sigma, 1 / sigma, trainable=False)
+
+        self.net = build(theta_dim + x_dim + 2, theta_dim, **kwargs)
```
[the remark this answers is not in the document]
Comment (francois-rozet) [reply lampe/inference.py]: I added an argument for the size of the time embedding.

File: lampe/inference.py (hunk 1 above)
Comment (glouppe) [reply lampe/inference.py:650]: Gotcha, I missed the `(rescaled)` in the docstring.

File: lampe/inference.py [hunk 6]
```diff
@@ -532,6 +548,177 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
...
+
+    def forward(self, theta: Tensor, x: Tensor, t: Tensor) -> Tensor:
+        r"""
+        Arguments:
+            theta: The parameters :math:`\theta`, with shape :math:`(*, D)`.
+            x: The observation :math:`x`, with shape :math:`(*, L)`.
+            t: The time :math:`t`, with shape :math:`(*,).`
+
+        Returns:
+            The score :math:`s_\phi(\theta, x, t)`, with shape :math:`(*,)`.
```
Comment (glouppe) [inline lampe/inference.py:631]: To avoid future confusion, it should be highlighted here that `s_phi` is not the score, but the rescaled score.

File: lampe/inference.py [hunk 7]
```diff
@@ -524,14 +540,183 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
...
+    .. math:: \arg\min_\phi \mathbb{E}_{p(\theta, x) p(t) p(\theta_t | \theta)}
+        \Big[ \big\|
+            s_\phi(\theta_t, x, t) -
+            (1 - \alpha(t)) \nabla_{\! \theta_t} \log p(\theta_t | \theta)
+        \big\|_2^2 \Big] .
+
+    For log-density computation and sampling, we solve the probability flow ODE
+
+    .. math:: \mathrm{d} \theta = \left[ -\frac{1}{2} \beta(t) \theta -
+        \frac{1}{2} \beta(t) (1 + \alpha(t)) s_\phi(\theta, x, t) \right] \, \mathrm{d} t
```
[the remark this answers is not in the document]
Comment (glouppe) [reply lampe/inference.py:581]: ^ OK given that `s_phi` is the rescaled score.

File: lampe/inference.py [hunk 8]
```diff
@@ -532,6 +548,177 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
         return -log_prob.mean()
 
 
+class NSE(nn.Module):
+    r"""Creates a neural score estimation (NSE) regression network.
+
+    The principle of neural score estimation is to train a regression network
+    :math:`s_\phi(\theta, x, t)` to approximate the score of a variance preserving
```
Comment (glouppe) [inline lampe/inference.py:555]: Same here, this might mislead users into believing `s_phi` is the score.

File: lampe/inference.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (glouppe) [reply lampe/inference.py:556]: Given that Song 2021 introduces at least two different variance-preserving processes, don't you think it would make it less confusing to tell explicitly the users which variant this class implements? To me "sub-VP SDEs" could be used as is if we just refer to the paper for further details. This name should be known to researchers interested in diffusion-based generative models.

File: lampe/inference.py [hunk 9]
```diff
@@ -532,6 +548,177 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
...
+                f=lambda theta, t: self.ode(theta, x, t),
+                time=x.new_tensor(1.0),
+                phi=(x, *self.parameters()),
+            ),
+            base=DiagNormal(self.zeros, self.ones).expand(batch_shape),
+        )
+
+
+class NSELoss(nn.Module):
+    r"""Creates a module that calculates the loss :math:`l` of a NSE regressor
```
Comment (glouppe) [inline lampe/inference.py:682]: "the" loss is quite vague and slightly abusive (in the sense that there are many different losses that could be used to train a score network), can you refer to a particular equation in the paper?

File: lampe/inference.py (hunk 2 above)
Comment (glouppe) [reply lampe/inference.py:678]: Yeah, I was more worried about the lack of consistency between the interfaces of the classes than about the particular implementation of the `sample` interface.

**Comment by francois-rozet** [comment]
Most comments have been addressed in the last commits. The description of NSE was mostly re-written.

File: lampe/inference.py [hunk 10]
```diff
@@ -532,20 +546,198 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
...
+        batch_shape = x.shape[:-1]
+
+        return NormalizingFlow(
+            transform=FFJTransform(
+                f=lambda theta, t: self.ode(theta, x, t),
+                time=x.new_tensor(1.0),
+                phi=(x, *self.parameters()),
+            ),
+            base=DiagNormal(self.zeros, self.ones).expand(batch_shape),
+        )
```
Comment (glouppe) [inline lampe/inference.py:676]: After experimenting with this part of the code, it would be nice to have a unified API such that `nse.flow` can be passed directly to the coverage utils. This is currently not convenient because `nse.flow.log_prob` is a noisy estimate of the log-density.

**Comment by glouppe** [comment]
Thanks for the changes! I have experimented with the codebase and confirm that it works fine -- it produces good results on SLCP, and the coverage is reasonable. 

+1 for merging in its current state
